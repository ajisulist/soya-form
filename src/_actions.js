import {
  SET_ENABLED_STATE_ACTION_TYPE,
  SET_IS_SUBMITTING_ACTION_TYPE,
  SET_IS_VALIDATING_ACTION_TYPE,
  MERGE_FIELDS_ACTION_TYPE,
  SET_VALUE_ACTION_TYPE,
  SET_VALUES_ACTION_TYPE,
  SET_DEFAULT_VALUE_ACTION_TYPE,
  SET_DEFAULT_VALUES_ACTION_TYPE,
  CLEAR_FIELD_ACTION_TYPE,
  SET_ERRORS_ACTION_TYPE,
  ADD_ERRORS_ACTION_TYPE,
  CLEAR_ERRORS_ACTION_TYPE,
  CLEAR_FORM_ACTION_TYPE,
  ADD_LIST_ITEM_ACTION_TYPE,
  ADD_LIST_ITEM_WITH_VALUE_ACTION_TYPE,
  REMOVE_LIST_ITEM_ACTION_TYPE,
  REORDER_LIST_ITEM_ACTION_TYPE,
  REORDER_LIST_ITEM_DEC_ACTION_TYPE,
  REORDER_LIST_ITEM_INC_ACTION_TYPE,
} from './_constants';

// Simple field related actions.
export const setFormEnabledState = (formId, isEnabled) => ({
  type: SET_ENABLED_STATE_ACTION_TYPE,
  formId,
  isEnabled,
});

export const setFormIsSubmittingState = (formId, isSubmitting) => ({
  type: SET_IS_SUBMITTING_ACTION_TYPE,
  formId,
  isSubmitting,
});

export const setIsValidating = (formId, map) => ({
  type: SET_IS_VALIDATING_ACTION_TYPE,
  formId,
  map,
});

export const mergeFields = (formId, fields) => ({
  type: MERGE_FIELDS_ACTION_TYPE,
  formId,
  fields,
});

export const setValue = (formId, fieldName, value) => ({
  type: SET_VALUE_ACTION_TYPE,
  formId,
  fieldName,
  value,
});

export const setValues = (formId, values) => ({
  type: SET_VALUES_ACTION_TYPE,
  formId,
  values,
});

export const setDefaultValue = (formId, fieldName, value) => ({
  type: SET_DEFAULT_VALUE_ACTION_TYPE,
  formId,
  fieldName,
  value,
});

export const setDefaultValues = (formId, values) => ({
  type: SET_DEFAULT_VALUES_ACTION_TYPE,
  formId,
  values,
});

export const clearField = (formId, fieldName) => ({
  type: CLEAR_FIELD_ACTION_TYPE,
  formId,
  fieldName,
});

export const setErrorMessages = (formId, fieldName, errorMessages) => ({
  type: SET_ERRORS_ACTION_TYPE,
  formId,
  fieldName,
  errorMessages,
});

export const addErrorMessages = (formId, errorMessages) => ({
  type: ADD_ERRORS_ACTION_TYPE,
  formId,
  errorMessages,
});

export const clearErrorMessages = (formId, fieldNames) => ({
  type: CLEAR_ERRORS_ACTION_TYPE,
  formId,
  fields: fieldNames.map(fieldName => ({ fieldName })),
});

export const clear = (formId) => ({
  type: CLEAR_FORM_ACTION_TYPE,
  formId,
});

// Repeatable field related action.
export const addListItem = (formId, fieldName, minLength, maxLength) => ({
  type: ADD_LIST_ITEM_ACTION_TYPE,
  formId,
  fieldName,
  minLength,
  maxLength,
});

export const addListItemWithValue = (formId, fieldName, values) => ({
  type: ADD_LIST_ITEM_WITH_VALUE_ACTION_TYPE,
  formId,
  values: values.map(value => ({
    ...value,
    fieldName: fieldName.concat(value.fieldName),
  })),
});

export const removeListItem = (formId, fieldName, index) => ({
  type: REMOVE_LIST_ITEM_ACTION_TYPE,
  formId,
  fieldName: fieldName.concat(index),
});

export const reorderListItemInc = (formId, fieldName, index, amount = 1) => ({
  type: REORDER_LIST_ITEM_INC_ACTION_TYPE,
  formId,
  fieldName: fieldName.concat(index),
  amount,
});

export const reorderListItemDec = (formId, fieldName, index, amount = 1) => ({
  type: REORDER_LIST_ITEM_DEC_ACTION_TYPE,
  formId,
  fieldName: fieldName.concat(index),
  amount,
});

export const reorderListItem = (formId, fieldName, index, targetIndex) => ({
  type: REORDER_LIST_ITEM_ACTION_TYPE,
  formId,
  fieldName: fieldName.concat(index),
  targetIndex,
});
