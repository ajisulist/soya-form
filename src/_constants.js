export const STATE_NAME = '__soyaForm';
export const CONTEXT_NAME = '__soyaForm';

const ID = 'form';
export const SET_DEFAULT_VALUE_ACTION_TYPE = `${ID}.setDefaultValue`;
export const SET_DEFAULT_VALUES_ACTION_TYPE = `${ID}.setDefaultValues`;
export const SET_VALUE_ACTION_TYPE = `${ID}.setValue`;
export const SET_VALUES_ACTION_TYPE = `${ID}.setValues`;
export const CLEAR_FIELD_ACTION_TYPE = `${ID}.clearField`;
export const MERGE_FIELDS_ACTION_TYPE = `${ID}.mergeFields`;
export const SET_IS_VALIDATING_ACTION_TYPE = `${ID}.setIsValidating`;
export const SET_IS_SUBMITTING_ACTION_TYPE = `${ID}.setIsSubmitting`;
export const SET_ERRORS_ACTION_TYPE = `${ID}.setErrors`;
export const ADD_ERRORS_ACTION_TYPE = `${ID}.addErrors`;
export const SET_ENABLED_STATE_ACTION_TYPE = `${ID}.setEnabledState`;
export const CLEAR_FORM_ACTION_TYPE = `${ID}.clearForm`;
export const CLEAR_ERRORS_ACTION_TYPE = `${ID}.clearErrors`;

export const ADD_LIST_ITEM_ACTION_TYPE = `${ID}.addListItem`;
export const ADD_LIST_ITEM_WITH_VALUE_ACTION_TYPE = `${ID}.addListItemWithValue`;
export const REMOVE_LIST_ITEM_ACTION_TYPE = `${ID}.removeListItem`;
export const REORDER_LIST_ITEM_ACTION_TYPE = `${ID}.reorderListItem`;
export const REORDER_LIST_ITEM_INC_ACTION_TYPE = `${ID}.reorderItemInc`;
export const REORDER_LIST_ITEM_DEC_ACTION_TYPE = `${ID}.reorderItemDec`;
export const INIT_FORM = `${ID}.INIT_FORM`;
export const INIT_FIELD = `${ID}.INIT_FIELD`;
