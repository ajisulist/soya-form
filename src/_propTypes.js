import PropTypes from 'prop-types';

export const formShape = PropTypes.shape([
  'lockSubmission',
  'unlockSubmission',
  'disable',
  'enable',
  'disableField',
  'enableField',
  'disableFields',
  'enableFields',
  'setValue',
  'setValues',
  'setDefaultValue',
  'setDefaultValues',
  'setErrors',
  'addErrors',
  'clearForm',
  'clearErrors',
  'getFormId',
  'regField',
  'unregField',
  'validateAll',
  'submit',
].reduce((propTypes, name) => ({
  ...propTypes,
  [name]: PropTypes.func.isRequired,
}), {}));
