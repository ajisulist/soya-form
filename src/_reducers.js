import {
  SET_ENABLED_STATE_ACTION_TYPE,
  SET_IS_SUBMITTING_ACTION_TYPE,
  SET_IS_VALIDATING_ACTION_TYPE,
  MERGE_FIELDS_ACTION_TYPE,
  SET_VALUE_ACTION_TYPE,
  SET_VALUES_ACTION_TYPE,
  SET_DEFAULT_VALUE_ACTION_TYPE,
  SET_DEFAULT_VALUES_ACTION_TYPE,
  CLEAR_FIELD_ACTION_TYPE,
  SET_ERRORS_ACTION_TYPE,
  ADD_ERRORS_ACTION_TYPE,
  CLEAR_ERRORS_ACTION_TYPE,
  CLEAR_FORM_ACTION_TYPE,
  ADD_LIST_ITEM_ACTION_TYPE,
  ADD_LIST_ITEM_WITH_VALUE_ACTION_TYPE,
  REMOVE_LIST_ITEM_ACTION_TYPE,
  REORDER_LIST_ITEM_ACTION_TYPE,
  REORDER_LIST_ITEM_DEC_ACTION_TYPE,
  REORDER_LIST_ITEM_INC_ACTION_TYPE,
} from './_constants';

import {
  isArray,
  isField,
  getFieldNames,
  removeObject,
  updateArray,
  removeArray,
  moveArray,
} from './_utils';

export const initialState = {
  field: {
    errorMessages: [],
    isEnabled: true,
    isValidating: false,
    touched: false,
    value: null,
  },
  fields: {
    array: [],
    object: {},
  },
  form: {
    isEnabled: true,
    isSubmitting: false,
  },
  forms: {},
};

const update = (reducer) => (state, action, fieldName) => {
  const newState = reducer(state[fieldName], action);
  if (isArray(state)) {
    return updateArray(state, newState, fieldName);
  }
  return {
    ...state,
    [fieldName]: newState,
  };
};

const field = (state = initialState.field, action) => {
  switch (action.type) {
    case SET_IS_VALIDATING_ACTION_TYPE:
      return {
        ...state,
        isValidating: action.isValidating,
        touched: true,
      };
    case SET_DEFAULT_VALUE_ACTION_TYPE:
    case SET_DEFAULT_VALUES_ACTION_TYPE:
    case SET_VALUE_ACTION_TYPE:
    case SET_VALUES_ACTION_TYPE:
    case ADD_LIST_ITEM_WITH_VALUE_ACTION_TYPE:
      if (state.value === action.value) {
        return state;
      }
      return {
        ...state,
        errorMessages: [],
        isEnabled: true,
        isValidating: false,
        touched: true,
        value: action.value,
      };
    case SET_ERRORS_ACTION_TYPE:
      return {
        ...state,
        errorMessages: action.errorMessages,
      };
    case ADD_ERRORS_ACTION_TYPE:
      return {
        ...state,
        errorMessages: state.errorMessages.concat(action.errorMessages),
      };
    case MERGE_FIELDS_ACTION_TYPE:
      return {
        ...state,
        ...action.object,
      };
    case CLEAR_ERRORS_ACTION_TYPE:
      return {
        ...state,
        errorMessages: [],
      };
    default:
      return state;
  }
};

const addField = (state, action) => {
  if (action.maxLength && state.length + 1 > action.maxLength) {
    if (state.length === action.maxLength) {
      return state;
    }
    return state.slice(0, action.maxLength);
  }
  const newState = update(field)(state, action, state.length);
  if (action.minLength && newState.length < action.minLength) {
    return addField(newState, action);
  }
  return newState;
};
const updateField = (state, action) => update(field)(state, action, action.fieldName);

const createFields = (initialState) => (state = initialState, action) => {
  const fieldName = action.fieldName[0];
  if (action.fieldName.length > 1) {
    if (isField(state[fieldName])) {
      throw new Error(`Unable to perform "${action.type}" action for "${action.fieldNames}" field, "${fieldName}" is a field`);
    }
    return updateFields(state, {
      ...action,
      fieldName: action.fieldName.slice(1),
    }, fieldName);
  }

  if (!isArray(state)) {
    switch (action.type) {
      case ADD_LIST_ITEM_ACTION_TYPE:
      case ADD_LIST_ITEM_WITH_VALUE_ACTION_TYPE:
      case REMOVE_LIST_ITEM_ACTION_TYPE:
      case REORDER_LIST_ITEM_ACTION_TYPE:
      case REORDER_LIST_ITEM_DEC_ACTION_TYPE:
      case REORDER_LIST_ITEM_INC_ACTION_TYPE:
        throw new Error(`"${action.type}" action expected an array state, got ${typeof state}`);
    }
  } else if (typeof fieldName !== 'number') {
    throw new Error(`"${fieldName}" field expected an object state, got array`);
  }

  const newAction = {
    ...action,
    fieldName,
  };

  switch (newAction.type) {
    case CLEAR_FIELD_ACTION_TYPE:
      return removeObject(state, fieldName);
    case SET_IS_VALIDATING_ACTION_TYPE:
    case SET_VALUE_ACTION_TYPE:
    case SET_VALUES_ACTION_TYPE:
    case SET_ERRORS_ACTION_TYPE:
    case ADD_ERRORS_ACTION_TYPE:
    case MERGE_FIELDS_ACTION_TYPE:
    case CLEAR_ERRORS_ACTION_TYPE:
      return updateField(state, newAction);
    case SET_DEFAULT_VALUE_ACTION_TYPE:
    case SET_DEFAULT_VALUES_ACTION_TYPE:
      if (state[fieldName]) {
        return state;
      }
      return updateField(state, newAction);
    case REMOVE_LIST_ITEM_ACTION_TYPE:
      return removeArray(state, fieldName);
    case ADD_LIST_ITEM_ACTION_TYPE:
    case ADD_LIST_ITEM_WITH_VALUE_ACTION_TYPE:
      return addField(state, newAction);
    case REORDER_LIST_ITEM_DEC_ACTION_TYPE:
      return moveArray(state, fieldName, fieldName - newAction.amount);
    case REORDER_LIST_ITEM_INC_ACTION_TYPE:
      return moveArray(state, fieldName, fieldName + newAction.amount);
    case REORDER_LIST_ITEM_ACTION_TYPE:
      return moveArray(state, fieldName, newAction.targetIndex);
  }
};

const arrayFields = createFields(initialState.fields.array);
const objectFields = createFields(initialState.fields.object);

const fields = (state, action) => {
  const fieldNames = getFieldNames(action.fieldName);
  if (fieldNames.length === 0) {
    return state;
  }
  const fieldName = fieldNames[0];
  const reducer = typeof fieldName === 'number' ? arrayFields : objectFields;
  return reducer(state, {
    ...action,
    fieldName: fieldNames,
    fieldNames: action.fieldNames || fieldNames.join('.'),
  });
};

const updateFields = update(fields);
const reduceFields = (state, action, values) => ({
  ...state,
  fields: values.reduce((state, valueAction) => fields(state, {
    ...action,
    ...valueAction,
  }), state.fields),
});

const form = (state = initialState.form, action) => {
  switch (action.type) {
    case SET_ENABLED_STATE_ACTION_TYPE:
      return {
        ...state,
        isEnabled: action.isEnabled,
      };
    case SET_IS_VALIDATING_ACTION_TYPE:
      return {
        ...state,
        fields: Object.keys(action.map).reduce((state, fieldName) => (
          fields(state, {
            ...action,
            fieldName,
            isValidating: action.map[fieldName],
          })
        ), state.fields),
      };
    case SET_IS_SUBMITTING_ACTION_TYPE:
      return {
        ...state,
        isSubmitting: action.isSubmitting,
      };
    case CLEAR_FIELD_ACTION_TYPE:
      const newState = fields(state.fields, action);
      if (Object.keys(newState).length === 0) {
        return removeObject(state, 'fields');
      }
      return {
        ...state,
        fields: newState,
      };
    case SET_DEFAULT_VALUE_ACTION_TYPE:
    case SET_VALUE_ACTION_TYPE:
    case SET_ERRORS_ACTION_TYPE:
    case ADD_LIST_ITEM_ACTION_TYPE:
    case REMOVE_LIST_ITEM_ACTION_TYPE:
    case REORDER_LIST_ITEM_ACTION_TYPE:
    case REORDER_LIST_ITEM_DEC_ACTION_TYPE:
    case REORDER_LIST_ITEM_INC_ACTION_TYPE:
      return {
        ...state,
        fields: fields(state.fields, action),
      };
    case SET_DEFAULT_VALUES_ACTION_TYPE:
    case SET_VALUES_ACTION_TYPE:
    case ADD_LIST_ITEM_WITH_VALUE_ACTION_TYPE:
      return reduceFields(state, action, action.values);
    case ADD_ERRORS_ACTION_TYPE:
      return reduceFields(state, action, action.errorMessages);
    case CLEAR_ERRORS_ACTION_TYPE:
    case MERGE_FIELDS_ACTION_TYPE:
      return reduceFields(state, action, action.fields);
    case CLEAR_FORM_ACTION_TYPE:
      return {
        ...removeObject(state, 'fields'),
        isEnabled: true,
      };
  }
};

export default (state = initialState.forms, action) => {
  switch (action.type) {
    case SET_ENABLED_STATE_ACTION_TYPE:
    case SET_IS_VALIDATING_ACTION_TYPE:
    case SET_IS_SUBMITTING_ACTION_TYPE:
    case SET_DEFAULT_VALUE_ACTION_TYPE:
    case SET_DEFAULT_VALUES_ACTION_TYPE:
    case SET_VALUE_ACTION_TYPE:
    case SET_VALUES_ACTION_TYPE:
    case CLEAR_FIELD_ACTION_TYPE:
    case SET_ERRORS_ACTION_TYPE:
    case ADD_ERRORS_ACTION_TYPE:
    case MERGE_FIELDS_ACTION_TYPE:
    case CLEAR_FORM_ACTION_TYPE:
    case CLEAR_ERRORS_ACTION_TYPE:
    case ADD_LIST_ITEM_ACTION_TYPE:
    case ADD_LIST_ITEM_WITH_VALUE_ACTION_TYPE:
    case REMOVE_LIST_ITEM_ACTION_TYPE:
    case REORDER_LIST_ITEM_ACTION_TYPE:
    case REORDER_LIST_ITEM_DEC_ACTION_TYPE:
    case REORDER_LIST_ITEM_INC_ACTION_TYPE:
      return {
        ...state,
        [action.formId]: form(state[action.formId], action),
      };
    default:
      return state;
  }
};
