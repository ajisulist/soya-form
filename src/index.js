import createForm from './createForm';
import createField from './createField';
import createRepeatable from './createRepeatable';
import withForm from './withForm';

export default {
  createForm,
  createField,
  createRepeatable,
  withForm,
};
