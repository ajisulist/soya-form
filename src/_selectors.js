import { STATE_NAME } from './_constants';
import { isArray, isField } from './_utils';

const _getState = state => state[STATE_NAME];

const _getForm = (state, formId) => () => _getState(state)[formId] || {};

const _getValues = field => {
  if (!field) {
    return null;
  }
  if (isField(field)) {
    return field.value;
  }
  let values = {};
  if (isArray(field)) {
    values = [];
  }
  for (const name in field) {
    values[name] = _getValues(field[name]);
  }
  return values;
};

const _hasErrors = field => {
  if (!field) {
    return false;
  }
  if (field.errorMessages) {
    return field.errorMessages.length > 0;
  }
  if (isArray(field)) {
    return field.some(field => _hasErrors(field));
  }
  const fieldNames = Object.keys(field);
  return fieldNames.some(name => _hasErrors(field[name]));
};

export default (state, formId) => {
  const getForm = _getForm(state, formId);
  const getFields = () => getForm().fields || null;

  const getField = fieldNames => {
    let field = getFields();
    if (!field) {
      return null;
    }
    for (const fieldName of fieldNames) {
      field = field[fieldName];
      if (!field) {
        return null;
      }
    }
    return field;
  };

  return {
    getValues: () => _getValues(getFields()),
    getField,
    getFieldValue: fieldNames => {
      const field = getField(fieldNames);
      return field && field.value;
    },
    getForm,
    getLength: fieldNames => {
      const field = getField(fieldNames);
      if (!field) {
        return 0;
      }
      return field.length;
    },
    hasErrors: () => _hasErrors(getFields()),
    isEnabled: () => getForm().isEnabled || false,
    isSubmitting: () => getForm().isSubmitting || false,
  };
};
