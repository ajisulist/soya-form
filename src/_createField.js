import React from 'react';
import PropTypes from 'prop-types';
import {
  CONTEXT_NAME,
} from './_constants';
import { formShape } from './_propTypes';
import {
  getDisplayName,
} from './_utils';

export default (Component) => {
  class CreateField extends React.Component {
    static displayName = getDisplayName('CreateField', Component);

    static propTypes = {
      form: formShape.isRequired,
      name: PropTypes.arrayOf(PropTypes.string).isRequired,
      isFormEnabled: PropTypes.bool.isRequired,
      isFormSubmitting: PropTypes.bool.isRequired,
      isDisabled: PropTypes.bool.isRequired,
      errorMessages: PropTypes.array.isRequired,
      isFieldEnabled: PropTypes.bool.isRequired,
      isValidating: PropTypes.bool.isRequired,
      touched: PropTypes.bool.isRequired,
      value: PropTypes.any,
      changeHandlers: PropTypes.arrayOf(PropTypes.func),
      changeValidators: PropTypes.arrayOf(PropTypes.func),
      createHandleChange: PropTypes.func.isRequired,
      asyncValidators: PropTypes.arrayOf(PropTypes.func),
      createHandleAsyncValidation: PropTypes.func.isRequired,
      submitValidators: PropTypes.arrayOf(PropTypes.func),
      createHandleValidateAll: PropTypes.func.isRequired,
    };

    static defaultProps = {
      changeHandlers: [],
      changeValidators: [],
      asyncValidators: [],
      submitValidators: [],
    };

    constructor(props, context) {
      super(props, context);
      this.__form = props.form;
      this.__asyncValidators = [];
      this.__changeValidators = [];
      this.__submitValidators = [];
    }

    componentWillMount() {
      const handleValidateAll = this.props.createHandleValidateAll([
        ...this.props.changeValidators,
        ...this.__changeValidators,
      ], [
        ...this.props.asyncValidators,
        ...this.__asyncValidators,
        ...this.props.submitValidators,
        ...this.__submitValidators,
      ]);
      this.__form.regField(this.props.name, handleValidateAll);
    }

    componentWillUnmount() {
      this.__form.unregField(this.props.name);
    }

    registerAsyncValidators(asyncValidators) {
      this.__asyncValidators = this.__asyncValidators.concat(asyncValidators);
    }

    registerChangeValidators(changeValidators) {
      this.__changeValidators = this.__changeValidators.concat(changeValidators);
    }

    registerSubmitValidators(submitValidators) {
      this.__submitValidators = this.__submitValidators.concat(submitValidators);
    }

    render() {
      const {
        createHandleChange,
        createHandleAsyncValidation,
        changeValidators,
        asyncValidators,
        ...props,
      } = this.props;
      delete props.changeHandlers;
      delete props.createHandleValidateAll;
      delete props.submitValidators;

      const validators = [
        ...changeValidators,
        ...this.__changeValidators,
      ];
      props.handleChange = createHandleChange(validators);
      props.handleAsyncValidation = createHandleAsyncValidation(validators, [
        ...asyncValidators,
        ...this.__asyncValidators,
      ]);
      props.registerChangeValidators = this.registerChangeValidators;
      props.registerAsyncValidators = this.registerAsyncValidators;
      props.registerSubmitValidators = this.registerSubmitValidators;

      return <Component {...props} />;
    }
  }

  return CreateField;
};
