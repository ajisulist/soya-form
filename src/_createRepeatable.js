import React from 'react';
import PropTypes from 'prop-types';
import { formShape } from './_propTypes';
import {
  createGetName,
  getDisplayName,
} from './_utils';

export default (Component) => {
  class CreateRepeatable extends React.Component {
    static displayName = getDisplayName('CreateRepeatable', Component);

    static propTypes = {
      form: formShape.isRequired,
      length: PropTypes.number.isRequired,
      maxLength: PropTypes.number,
      minLength: PropTypes.number,
      name: PropTypes.arrayOf(PropTypes.string).isRequired,
      addListItem: PropTypes.func.isRequired,
      removeListItem: PropTypes.func.isRequired,
      reorderListItem: PropTypes.func.isRequired,
      reorderListItemDec: PropTypes.func.isRequired,
      reorderListItemInc: PropTypes.func.isRequired,
    };

    static defaultProps = {
      minLength: 0,
    };

    addListItem = () => this.props.addListItem(this.props.name)();

    removeListItem = (index) => this.props.removeListItem(this.props.name)(index);

    reorderListItem = (index, targetIndex) => this.props.reorderListItem(this.props.name)(index, targetIndex);

    reorderListItemDec = (index, amount) => this.props.reorderListItemDec(this.props.name)(index, amount);

    reorderListItemInc = (index, amount) => this.props.reorderListItemInc(this.props.name)(index, amount);

    render() {
      return (
        <div>
          {new Array(this.props.length).fill(null).map((_, index) => (
            <Component
              {...this.props}
              index={index}
              key={index}
              name={createGetName(this.props.name, index)}
              totalItem={this.props.length}
              addListItem={this.addListItem}
              removeListItem={this.removeListItem}
              reorderListItem={this.reorderListItem}
              reorderListItemDec={this.reorderListItemDec}
              reorderListItemInc={this.reorderListItemInc}
            />
          ))}
        </div>
      );
    }
  }

  return CreateRepeatable;
};
