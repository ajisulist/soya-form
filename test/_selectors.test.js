import { STATE_NAME } from '../src/_constants';
import createSelector from '../src/_selectors';

describe('simpleFormSelector', () => {
  const simpleFormId = 'simpleForm';
  const uninitializedFormId = 'uninitializedForm';
  const state = {
    [STATE_NAME]: {
      [simpleFormId]: {
        isEnabled: false,
        isSubmitting: false,
        fields: {
          foo: {
            bar: [
              {
                errorMessages: [],
                isEnabled: true,
                isValidating: false,
                touched: false,
                value: 'value 1',
              },
              {
                errorMessages: [],
                isEnabled: true,
                isValidating: false,
                touched: false,
                value: 'value 2',
              },
            ],
          },
        },
      },
      [uninitializedFormId]: {},
    },
  };
  const simpleFormSelector = createSelector(state, simpleFormId);
  const uninitializedFormSelector = createSelector(state, uninitializedFormId);
  const unknownFormSelector = createSelector(state, 'unknownForm');

  it('should get the values', () => {
    expect(uninitializedFormSelector.getValues()).toMatchSnapshot();
    expect(simpleFormSelector.getValues()).toMatchSnapshot();
  });

  it('should get the field', () => {
    expect(simpleFormSelector.getField(['foo', 'bar'])).toMatchSnapshot();
    expect(simpleFormSelector.getField(['foo', 'oof'])).toMatchSnapshot();
    expect(uninitializedFormSelector.getField(['foo', 'bar'])).toMatchSnapshot();
  });

  it('should get the field value', () => {
    expect(simpleFormSelector.getFieldValue(['foo', 'bar', 0])).toMatchSnapshot();
  });

  it('should get the form', () => {
    expect(unknownFormSelector.getForm()).toMatchSnapshot();
    expect(simpleFormSelector.getForm()).toMatchSnapshot();
  });

  it('should get the field length', () => {
    expect(simpleFormSelector.getLength(['foo', 'bar'])).toMatchSnapshot();
    expect(uninitializedFormSelector.getLength(['foo', 'bar'])).toMatchSnapshot();
  });

  it('should show whether the form has errors or not', () => {
    expect(uninitializedFormSelector.hasErrors()).toMatchSnapshot();
    expect(simpleFormSelector.hasErrors()).toMatchSnapshot();
  });

  it('should show whether the form is enabled or not', () => {
    expect(simpleFormSelector.isEnabled()).toMatchSnapshot();
  });

  it('should show whether the form is locking submission or not', () => {
    expect(simpleFormSelector.isSubmitting()).toMatchSnapshot();
  });
});
