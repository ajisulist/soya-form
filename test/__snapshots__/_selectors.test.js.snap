// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simpleFormSelector should get the field 1`] = `
Array [
  Object {
    "errorMessages": Array [],
    "isEnabled": true,
    "isValidating": false,
    "touched": false,
    "value": "value 1",
  },
  Object {
    "errorMessages": Array [],
    "isEnabled": true,
    "isValidating": false,
    "touched": false,
    "value": "value 2",
  },
]
`;

exports[`simpleFormSelector should get the field 2`] = `undefined`;

exports[`simpleFormSelector should get the field 3`] = `null`;

exports[`simpleFormSelector should get the field length 1`] = `2`;

exports[`simpleFormSelector should get the field length 2`] = `0`;

exports[`simpleFormSelector should get the field value 1`] = `"value 1"`;

exports[`simpleFormSelector should get the form 1`] = `Object {}`;

exports[`simpleFormSelector should get the form 2`] = `
Object {
  "fields": Object {
    "foo": Object {
      "bar": Array [
        Object {
          "errorMessages": Array [],
          "isEnabled": true,
          "isValidating": false,
          "touched": false,
          "value": "value 1",
        },
        Object {
          "errorMessages": Array [],
          "isEnabled": true,
          "isValidating": false,
          "touched": false,
          "value": "value 2",
        },
      ],
    },
  },
  "isEnabled": false,
  "isSubmitting": false,
}
`;

exports[`simpleFormSelector should get the values 1`] = `null`;

exports[`simpleFormSelector should get the values 2`] = `
Object {
  "foo": Object {
    "bar": Array [
      "value 1",
      "value 2",
    ],
  },
}
`;

exports[`simpleFormSelector should show whether the form has errors or not 1`] = `false`;

exports[`simpleFormSelector should show whether the form has errors or not 2`] = `false`;

exports[`simpleFormSelector should show whether the form is enabled or not 1`] = `false`;

exports[`simpleFormSelector should show whether the form is locking submission or not 1`] = `false`;
